analyses/cnv-chrom-plot/util/bin-coverage.R

```diff
@@ -0,0 +1,142 @@
+  # Given a sample_id, CN segment ranges, and binned genome ranges object, 
+  # make a call for each bin on what CN copy status has the most coverage in the bin. 
+  # Uses bp_per_bin function. 
+  #
+  # Args:
+  #   sample_id: A string that corresponds to a single biospecimen id
+  #   seg_ranges: A GenomicRanges object that contains a `status` and a `biospecimen` slot.  
+  #               The `biospecimen slot will be used to split out the `sample_id`'s corresponding ranges.   
+  #               The `status` slot should have gain/loss/neutral. 
+  #   bin_ranges: A binned GenomicRanges made from tileGenome that has been uncompressed with `unlist`. 
+  #   perc_delta_threshold: What covereage percent difference do you need to make the call?
+  #
+  # Returns:
+  #  a small data.frame that contains the status call of the sample for each bin. 
+  #
+  # Extract the ranges for this sample
+  sample_seg_ranges <- seg_ranges[which(seg_ranges$biospecimen == sample_id)]
+  
+  # Split ranges into their respective statuses
+  gain_ranges <- sample_seg_ranges[sample_seg_ranges$status == "gain"]
+  loss_ranges <- sample_seg_ranges[sample_seg_ranges$status == "loss"]
+  neutral_ranges <- sample_seg_ranges[sample_seg_ranges$status == "neutral"]
+  
+  # Calculate length of each type of status per bin
+  gain_per_bin <- bp_per_bin(bin_ranges, gain_ranges)
+  loss_per_bin <- bp_per_bin(bin_ranges, loss_ranges)
+  neutral_per_bin <- bp_per_bin(bin_ranges, neutral_ranges)
+  
+  # Format this data into one data.frame where each row is a bin
+  bin_bp_status <- data.frame(
+    bin = as.numeric(1:length(bin_ranges)),
+    # Keep bin width
+    bin_width = bin_ranges@ranges@width
+  ) %>%
+    # Join gains coverage data
+    dplyr::left_join(gain_per_bin,
+                     by = "bin"
+    ) %>%
+    # Join loss coverage data
+    dplyr::left_join(loss_per_bin,
+                     by = "bin",
+                     suffix = c(".gain", ".loss")
+    ) %>%
+    # Join neutral coverage data
+    dplyr::left_join(neutral_per_bin,
+                     by = "bin"
+    ) %>%
+    # If there is an NA, at this point we can assume it means 0
+    dplyr::mutate_at(
+      dplyr::vars(
+        dplyr::starts_with("bp_per_bin")
+      ),
+      ~ tidyr::replace_na(., 0)
+    ) %>%
+    # Reformat neutral so it is like the others
+    dplyr::rename(bp_per_bin.neutral = bp_per_bin) %>%
+    # Calculate the bins percentage of each status
+    dplyr::mutate(
+      perc_gain = bp_per_bin.gain / bin_width,
+      perc_loss = bp_per_bin.loss / bin_width,
+      perc_neutral = bp_per_bin.neutral / bin_width
+    ) %>%
+    # Use these percentages for declaring final call per bin based on
+    # the perc_delta_threshold
+    dplyr::mutate(
+      status = dplyr::case_when(
+        (perc_gain - perc_loss) > perc_delta_threshold ~ "gain",
+        (perc_loss - perc_gain) > perc_delta_threshold ~ "loss",
+        TRUE ~ "neutral"
```

Comment:
I do think we should probably have some threshold for neutral as well. If a region is half gain and half loss, we probably don't want to code it as `neutral`. In other contexts (#644 and other CN analysis) we have been using an `unstable` call, which seems like it could be applied here too.

Reply:
So perhaps `frac_gain` + `frac_loss` is above a certain threshold, we'll call it `unstable`? What's a good number? Lol

Reply:
I think the logic we used was to set a pretty high threshold (no lower than ~60%), then use the same absolute threshold for all, so something like
```
frac_gain > threshold ~ "gain",
frac_loss > threshold ~ "loss",
frac_gain + frac_loss > threshold ~ "unstable",
frac_neutral > threshold  ~ "neutral",
TRUE ~ NA 
```

Reply:
Currently, I have `NA`s for when a region is uncallable. In this instance how do we distinguish these `NA`s from `uncallable` `NA`s?

Reply:
I mean, in some sense we are saying we can't make a call with the NA here, but I see the point. Part of the thought here was to get some empirical measure of how often we don't meet any particular threshold. Maybe the right thing to do is say "unstable" if `frac_neutral < 0.5`?

Reply:
I've added uncallable_ranges into the same function so we can use logic all at once. How is this:
```    dplyr::mutate(
      status = dplyr::case_when(
        frac_uncallable > uncallable_threshold ~ NA
        frac_gain > threshold ~ "gain",
        frac_loss > threshold ~ "loss",
        frac_gain + frac_loss > threshold ~ "unstable",
        TRUE  ~ "neutral",
      )
```

The order is important here, in that if the uncallable_threshold is met, we don't want to make any call besides `NA` but if the other items are called, we call it `neutral`?

Reply:
I like that, but is this crazy?
```
  status = dplyr::case_when(
        frac_uncallable > uncallable_threshold ~ NA
        frac_gain > threshold ~ "gain",
        frac_loss > threshold ~ "loss",
        frac_neutral > threshold ~ "neutral",
        TRUE ~ "unstable"
      )

Reply:
Turns out `dplyr::case_when` doesn't like `NA`s.

Reply:
But otherwise, I can do this strategy.

Reply:
Does it not like NA, or did it not like my missing comma?

Reply:
I put back the missing comma. It was the NA it doesn't like.

Reply:
Weird. That seems like a flaw in `case_when()`. They probably have reasons, and it is not too hard to work around, but I might post about it.

Reply:
With everything else being the same, if I put an NA as an assignment, it rerturns with: ```
Error: must be a logical vector, not a character vector
Call `rlang::last_error()` to see a backtrace 
```
and backtrace says
```
  1. base::data.frame(bin = as.numeric(1:length(bin_ranges)), bin_width = bin_ranges@ranges@width)
 15. dplyr::left_join(., gain_per_bin, by = "bin")
 16. dplyr::rename(., bp_per_bin.gain = bp_per_bin)
 17. dplyr::left_join(., loss_per_bin, by = "bin")
 18. dplyr::rename(., bp_per_bin.loss = bp_per_bin)
  1. dplyr::left_join(., neutral_per_bin, by = "bin")
 15. dplyr::rename(., bp_per_bin.neutral = bp_per_bin)
 16. dplyr::left_join(., uncallable_per_bin, by = "bin")
 17. dplyr::rename(., bp_per_bin.uncallable = bp_per_bin)
 18. dplyr::mutate_at(...)
  1. dplyr::mutate(...)
 15. dplyr::mutate(...)
 16. dplyr:::replace_with(...)
 17. dplyr:::check_type(val, x, name)
 18. dplyr:::glubort(header, "must be {friendly_type_of(template)}, not {friendly_type_of(x)}")
```

Reply:
I have (irresponsibly) little context but everything else is a character (`"gain"`) - does an `NA` here need to be [`NA_character_`](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/NA)? This is the opposite of what I would expect with the error message, but I feel that I've run into something similar recently.

Reply:
Jackie is correct. It needs to be the correct NA. Which seems dumb, but there you are. 
I tweeted about it and came to the same conclusion as I did so: https://twitter.com/jashapiro/status/1248229154648199168?s=20

Reply:
I'll never need context again 😂

Reply:
This is good information about `case_when()`'s behavior for the future, but I think in this instance, we'll stick with calling everything explicitly (so as is).